### PR TITLE
PE-4420: implement-multiple-downloads-for-files

### PR DESCRIPTION
**Pull request by kunstmusik**
Implements multi-file downloading for files on public and private drives with ability to skip, retry, and cancel in download error scenarios. 

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/13kdqniv14mag

### CODE REVIEW

File: lib/download/limits.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+final privateDownloadMobileSizeLimit = const MiB(300).size;
+
+Future<int> calcDownloadSizeLimit(bool isPublic,
+    {DeviceInfoPlugin? deviceInfo}) async {
+  if (isPublic) {
+    if (AppPlatform.getPlatform() == SystemPlatform.Web) {
+      final info = await (deviceInfo ?? DeviceInfoPlugin()).deviceInfo;
+
+      final isFirefox =
+          info is WebBrowserInfo && info.browserName == BrowserName.firefox;
```
Comment (thiagocarvalhodev) [inline lib/download/limits.dart:24]: Consider adding a helper method or property to the AppPlatform to abstract the logic to check the browser.

Something like:
```dart
return AppPlatform.isFirefox()
          ? publicDownloadFirefoxSizeLimit
          : publicDownloadWebSizeLimit;

File: lib/download/limits.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      return publicDownloadUnknownPlatformSizeLimit;
+    }
+  } else {
+    if (AppPlatform.getPlatform() == SystemPlatform.Web) {
+      final info = await (deviceInfo ?? DeviceInfoPlugin()).deviceInfo;
+      final isFirefox =
+          info is WebBrowserInfo && info.browserName == BrowserName.firefox;
+      return isFirefox
+          ? privateDownloadFirefoxSizeLimit
+          : privateDownloadWebSizeLimit;
```
Comment (thiagocarvalhodev) [inline lib/download/limits.dart:40]: We are repeating almost the same code of the previous if. We can also split this function into smaller functions. 

Consider this refactor:

```dart
Future<int> calcDownloadSizeLimit(bool isPublic,
    {DeviceInfoPlugin? deviceInfo}) async {
  if (AppPlatform.getPlatform() == SystemPlatform.Web) {
    return _webDownloadLimit(isPublic, deviceInfo);
  } else if (AppPlatform.isMobile) {
    return _mobileLimit(isPublic);
  }

  return _unknownPlatformLimit(isPublic);
}

Future<int> _webDownloadLimit(
    bool isPublic, DeviceInfoPlugin? deviceInfo) async {
  final info = await (deviceInfo ?? DeviceInfoPlugin()).deviceInfo;

  final isFirefox =
      info is WebBrowserInfo && info.browserName == BrowserName.firefox;

  if (isPublic) {
    return isFirefox
        ? publicDownloadFirefoxSizeLimit
        : publicDownloadWebSizeLimit;
  }

  return isFirefox
      ? privateDownloadFirefoxSizeLimit
      : privateDownloadWebSizeLimit;
}

int _mobileLimit(bool isPublic) {
  if (isPublic) {
    return publicDownloadMobileSizeLimit;
  }

  return publicDownloadMobileSizeLimit;
}

int _unknownPlatformLimit(bool isPublic) {
  if (isPublic) {
    return publicDownloadUnknownPlatformSizeLimit;
  }

  return publicDownloadUnknownPlatformSizeLimit;
}

File: lib/download/multiple_file_download_modal.dart [hunk 3]
```diff
@@ -60,53 +97,183 @@ class MultipleFilesDonwload extends StatelessWidget {
...
+                  appLocalizationsOf(context)
+                      .fileFailedToDownloadFileAbovePublicLimit,
+                );
+                break;
+              case FileDownloadFailureReason.fileNotFound:
+                content =
+                    Text(appLocalizationsOf(context).tryAgainDownloadingFile);
+                break;
+              case FileDownloadFailureReason.networkConnectionError:
+                // TODO: Localize text
```
Comment (thiagocarvalhodev) [inline lib/download/multiple_file_download_modal.dart:242]: @kunstmusik , I'll help you with the localizely

File: lib/download/multiple_download_bloc.dart [hunk 4]
```diff
@@ -1,81 +1,237 @@
...
+
+  Future<void> startDownload(
+      StartDownload event, Emitter<MultipleDownloadState> emit) async {
+    items = event.items;
+    skippedFiles.clear();
+
+    // check all files from same drive
+    var firstFile = items[0];
+
+    // Currently assumes all files from the same drive. May need code like below
```
Comment (thiagocarvalhodev) [inline lib/download/multiple_download_bloc.dart:81]: That's a good point, although now it's not possible to do it. We can remove that verification

File: lib/download/multiple_download_bloc.dart [hunk 5]
```diff
@@ -1,81 +1,237 @@
...
+  bool canceled = false;
+  int currentFileIndex = 0;
+  List<ARFSFileEntity> skippedFiles = [];
+  SecretKey? driveKey;
+
+  late ARFSDriveEntity drive;
+  late ZipEncoder zipEncoder;
+  late OutputStream outputStream;
+  late List<ARFSFileEntity> items;
+  late String outFileName;
```
Comment (thiagocarvalhodev) [inline lib/download/multiple_download_bloc.dart:41]: Do we need to expose these properties? If not, consider making them private

File: lib/download/multiple_download_bloc.dart [hunk 6]
```diff
@@ -1,81 +1,237 @@
...
+    }
+
+    _initializeEncoder();
+    outFileName =
+        event.folderName != null ? '${event.folderName}.zip' : 'Archive.zip';
+
+    await _downloadMultipleFiles(emit);
+  }
+
+  Future<void> resumeDownload(
```
Comment (thiagocarvalhodev) [inline lib/download/multiple_download_bloc.dart:122]: We may not need to expose these methods. Consider implementing them private if possible

**Review by thiagocarvalhodev — COMMENTED** [review]
Great job, @kunstmusik !! I didn't test the feature, I'll do it in the next couple of hours. So far, I'm just leaving a couple of suggestions.

File: lib/download/limits.dart (hunk 1 above)
Comment (kunstmusik) [reply lib/download/limits.dart]: Added to AppPlatform, had to make async so calling code uses `await AppPlatform.isFirefox()`.

File: lib/download/limits.dart (hunk 2 above)
Comment (kunstmusik) [reply lib/download/limits.dart]: Applied refactor, using ternary operators for most of the helper functions.

File: lib/download/multiple_download_bloc.dart (hunk 4 above)
Comment (kunstmusik) [reply lib/download/multiple_download_bloc.dart]: Removed the commented out test code, but also added a sanity check that the items list is not empty.

File: lib/download/multiple_download_bloc.dart (hunk 5 above)
Comment (kunstmusik) [reply lib/download/multiple_download_bloc.dart]: Renamed to make private, thanks!

**Review by thiagocarvalhodev — APPROVED** [review]
LGTM

**Review by thiagocarvalhodev — APPROVED** [review]
LGTM!

**Pull request merged**